metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala

```diff
@@ -0,0 +1,222 @@
+      sourceFiles: Set[Path],
+      sourceDirectories: Set[Path],
+      semanticdDirectories: Set[Path]
+  )
+
+  private def collectFilesToWatch(buildTargets: BuildTargets): FilesToWatch = {
+    val sourceDirectoriesToWatch = mutable.Set.empty[Path]
+    val sourceFilesToWatch = mutable.Set.empty[Path]
+
+    def collect(path: AbsolutePath): Unit = {
+      if (buildTargets.isInsideSourceRoot(path)) {
+        () // Do nothing, already covered by a source root
+      } else if (path.isScalaOrJava) {
+        sourceFilesToWatch.add(path.toNIO)
+      } else {
+        sourceDirectoriesToWatch.add(path.toNIO)
+      }
+    }
+    // Watch the source directories for "goto definition" index.
+    buildTargets.sourceRoots.foreach(collect)
+    buildTargets.sourceItems.foreach(collect)
+    val semanticdbs = buildTargets.scalacOptions.flatMap { item =>
+      for {
+        scalaInfo <- buildTargets.scalaInfo(item.getTarget)
+        targetroot = item.targetroot(scalaInfo.getScalaVersion)
+        path = targetroot.resolve(Directories.semanticdb) if !targetroot.isJar
+      } yield path.toNIO
+    }
+
+    FilesToWatch(
+      sourceFilesToWatch.toSet,
+      sourceDirectoriesToWatch.toSet,
+      semanticdbs.toSet
+    )
+  }
+
+  /**
+   * Start file watching
+   *
+   * Contains platform specific file watch initialization logic
+   *
+   * @param workspace current project workspace directory
+   * @param filesToWatch source files and directories to watch
+   * @param callback to execute on FileWatchEvent
+   * @param watchFilter predicate that filters which files
+   *        generate a FileWatchEvent on create/delete/change
+   * @return a dispose action resources used by file watching
+   */
+  private def startWatch(
+      workspace: Path,
+      filesToWatch: FilesToWatch,
+      callback: FileWatcherEvent => Unit,
+      watchFilter: Path => Boolean
+  ): () => Unit = {
+    if (scala.util.Properties.isMac) {
+      // Due to a hard limit on the number of FSEvents streams that can be opened on macOS,
+      // only the root workspace directory is registered for a recursive watch.
+      // However, the events are then filtered to receive only relevant events
+      // and also to hash only revelevant files when watching for changes
+
+      val trie = PathTrie(
+        filesToWatch.sourceFiles ++ filesToWatch.sourceDirectories ++ filesToWatch.semanticdDirectories
+      )
+      val isWatched = trie.containsPrefixOf _
+
+      val repo = initFileTreeRepository(
+        path => watchFilter(path) && isWatched(path),
+        callback
+      )
+      repo.register(workspace, Int.MaxValue)
+      () => repo.close()
+    } else {
+      // Other OSes register all the files and directories individually
+      val repo = initFileTreeRepository(watchFilter, callback)
+
+      // TODO(@pvid) swoval's FileTreeRepository should be able to create watch
+      // for files/directories that do not exist yet. However, there is an issue
```

Comment:
If I'm reading this correctly, you're saying that on linux or windows that the semanticdb file changes are not detected if the swoval watch is created before the semantic db file parent directories exist?

I'm not sure what is going on since there are test cases in swoval for the use case of watching a file when its parent doesn't exist, but I believe you that it is not working.

Reply:
I scanned the PR at a high level and it does look pretty good to me overall.

Reply:
unfortunately no - semanticdb file changes are not detected if the _source_ directories are not created before registering the watches. It makes no sense to me. 

There might be some subtle bug when registering multiple non-existing files, but it eludes me - I did not manage to come up with a failing swoval test case

Reply:
Huh, that is bizarre. But unfortunately it has definitely been my experience that some file watching bugs are very difficult to reproduce (which your inability to create a test case sadly confirms).